### PR TITLE
feat: Move all object configurations to dagger modules

### LINKED ISSUE #174: Add all object instantiation to Dagger modules
Change all configurable objects to dependency injectable components which can be easily swapped in and out for testing

### PR DESCRIPTION
**Pull request by iamareebjamal**
Fixes #174 

- Refactor modules
- Add barcode module
- Scan all components to look for any dependency left to be included in dagger
- Include sub-modules in modules to create structured dependency graph
- Update Google Vision API to solve a bug for projects using RxJava
- Change Leak Canary watch to appropriate lifecycle method to avoid false positives

### CODE REVIEW

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+public class BarcodeModule {
+
+    @Provides
+    @Singleton
+    @Named("barcodeEmitter")
+    PublishSubject<Notification<Barcode>> providesBarcodeEmitter() {
+        return PublishSubject.create();
+    }
+
+    @Provides
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Huge memory leak

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Plus, no one will be able to take selfies then ( you'll lock the camera forever if you make it a singleton )

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Detector.Processor<Barcode> providesProcessor(@Named("barcodeEmitter") PublishSubject<Notification<Barcode>> emitter) {
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: I am talking about `Detector.Processor<Barcode>` and `BarcodeDetector`.

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java (hunk 1 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Creating singleton is not a good practice in Java, but we need it for modules which will be used in several views throughout the app. Cameras and presenters are not such. They are mapped 1:1 to the views and also, they are bound to a very tight lifecycle. Thus, can't reside in memory forever

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java (hunk 2 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Processor can be. Detector can't. It depends on Context

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Holding context for time more than needed is memory leak

**Comment by mejariamol** [comment]
@iamareebjamal Whats wrong with travis ci build ?

**Comment by iamareebjamal** [comment]
Travis does not support latest repository. Have to fallback to previous version

**Comment by iamareebjamal** [comment]
@arp95 @mahikaw @Shailesh351 @shreyasnbhat @rohanagarwal94 Please review

**Comment by iamareebjamal** [comment]
@niranjan94 Please merge

**Pull request merged**